Use different ipg for the !ipg command

hgarus opened the pull request:
Fixes #35

hgarus committed:
Use different ipg for the !ipg command

Switch to the ipg used and maintained by MTG Familiar from the annotated
ipg on blogs.magicjudges.org. In addition to being somewhat less cluttered
it is also all on one page.

Download and parse the ipg once on startup and not on every query.

Seperate parsing and output-formatting into several different functions.

Add some tests for the ipg module.

hgarus committed:
Small fix to tests

hgarus committed:
Exclude test directory for duplication analysis

bra1n commented on modules/rules/ipg.js:
```diff
@@ -1,143 +1,266 @@
+    formatChapterEntry(entry) {
+        const text = entry.text || this.formatPreview(this.ipgData[entry.sections[0]]);
+
+        return `**IPG ${entry.title}**
```
Could we use `\n` here instead of literal line breaks?

hgarus replied:
I guess I wouldn't mind moving to something like
```javascript
return [
  `**IPG ${entry.title}**`,
  text,
  `**Available Sections**: ${this.formatSections(entry.sections)}`
].join('\n\n');
```

Is there a specific reason you dislike template literals with line breaks?

bra1n replied:
Yes, because it messes with the indentation. You'll either indent it to have it match the rest of the code (and then end up with extra spaces in your message) or not indent it to send the right amount of spaces, but end up with messy looking code. That's why I prefer the line break-joins or simply using `\n` instead of a literal line break.

bra1n commented on modules/rules/ipg.js:
```diff
@@ -1,143 +1,266 @@
+    formatSectionEntry(entry) {
+        const penalty = entry.penalty ? `\nPenalty: ${entry.penalty}\n` : '';
+        const text = entry.text || this.formatPreview(entry.subsectionContents[entry.subsections[0]]);
+        return `**IPG ${entry.title}**${penalty}
```
See Line 195 :)

bra1n commented on server.js:
```diff
@@ -18,7 +18,7 @@ const modules = [
 const handlers = {};
 modules.forEach(module => {
-    const moduleObject = new (require("./modules/" + module + '.js'))();
+    const moduleObject = new (require("./modules/" + module + '.js'))(true);
```
Suggestion: default initialize parameter to `true` and pass in `false` when running tests. I'd rather have the production call be without a parameter and the test call have the extra parameter.

bra1n commented on test/rules/ipg.js:
```diff
@@ -0,0 +1,256 @@
+            ipg = new IPG();
+            return ipg.init();
+        });
+        it("should have downloaded and parsed the ipg", function () {
```
Good test, but it relies on the availability of the online document, plus I couldn't run this test without internet connection. Maybe have a local copy of the IPG HTML for testing purposes available?

hgarus replied:
It feels like once we go there, we could just stop downloading it altogether and always use an integrated version. As long as the bot breaks if the online document is a missing I don't really mind the tests breaking as well.

I can see the argument about running the tests without an internet connection though.

bra1n reviewed (changes requested):
Wow, good job! I only have found very minor issues and I LOVE that you even took the time to add some tests!

bra1n commented on modules/rules/ipg.js:
```diff
@@ -1,143 +1,266 @@
-    constructor() {
-        this.location = "http://blogs.magicjudges.org/rules/ipg";
+    constructor(initialize) {
+        this.location = "https://sites.google.com/site/mtgfamiliar/rules/InfractionProcedureGuide-light.html";
```
I'd use a config variable for this, too, and default it to the above URL.

hgarus committed:
Various fixes for pull request

Call initialize by default on constructor calls.
Load the ipg from disk instead of the net during tests.
Make the download URL configurable through the environment.
Update README and help command.

hgarus committed:
Move away from multiline template literals
